ocaml/xenopsd/xc/device.ml

```diff
@@ -92,6 +90,170 @@ module Profile = struct
+    in
+    let cancel_watch =
+      Inotify.add_watch notifd
+        (Filename.dirname service.cancel_path)
+        [Inotify.S_Create]
+    in
+    Monitor.add pollfd notifd ;
+
+    let close_fds () =
+      Monitor.remove pollfd notifd ;
+      Polly.close pollfd ;
+      Inotify.rm_watch notifd ready_watch ;
+      Inotify.rm_watch notifd cancel_watch ;
+      Unix.close notifd
+    in
+
+    let wait ~task ~pid_path ~cancel_path ~for_s ~service_name =
+      let start_time = Mtime_clock.elapsed () in
+      let poll_period_ms = 1000 in
+      let cancel () =
+        (* create and empty file to trigger the watch and delete it
+           immediately *)
+        Unixext.atomic_write_to_file cancel_path 0o200 (fun fd ->
+            Unixext.really_write fd "" 0 0
+        ) ;
+        Unixext.unlink_safe cancel_path
+      in
+      let collect_watches (acc : watch_trigger) (event : Inotify.event) =
+        let file_watch = function
+          | _, [Inotify.Create], _, Some file
+            when file = Filename.basename pid_path ->
+              Finished
+          | _, [Inotify.Create], _, Some file
+            when file = Filename.basename cancel_path ->
+              Cancelled
+          | _ ->
+              Empty
+        in
+        match (acc, file_watch event) with
+        | Finished, _ | _, Finished ->
+            Finished
+        | Cancelled, _ | _, Cancelled ->
+            Cancelled
+        | Empty, Empty ->
+            Empty
+      in
+
+      let cancellable_watch () =
+        let event = ref Empty in
+        let rec poll_loop () =
+          try
+            ignore
+            @@ Polly.wait pollfd 1 poll_period_ms (fun _ fd events ->
+                   if Polly.Events.(test events inp) then
+                     event :=
+                       List.fold_left collect_watches !event (Inotify.read fd)
+               ) ;
+
+            let current_time = Mtime_clock.elapsed () in
+            let elapsed_time =
+              Mtime.Span.(to_s (abs_diff start_time current_time))
+            in
+
+            match !event with
+            | Empty when elapsed_time < for_s ->
+                poll_loop ()
+            | Finished ->
+                Ok ()
+            | Cancelled ->
+                Error (ECancelled task)
+            | Empty ->
+                let err_msg =
+                  if alive service_name then
+                    "Timeout reached while starting service"
+                  else
+                    "Service exited unexpectedly"
+                in
+                Error (Ioemu_failed (service_name, err_msg))
+          with e ->
+            let err_msg =
+              Printf.sprintf
+                "Exception while waiting for service %s to be ready: %s"
+                service_name (Printexc.to_string e)
+            in
+            Error (Failure err_msg)
+        in
+
+        Xenops_task.with_cancel task cancel poll_loop
+      in
+      cancellable_watch ()
+    in
```

Comment:
This is a lot simpler to do in the Lwt variant of inotify (see https://github.com/xapi-project/xen-api/pull/4670/commits/9d12dcd9ebbe8515606532d5c10b20c6591414a6), but probably because Lwt already takes care of wrapping the event loop and waiting for an event, whereas here you have to do it "manually".

Also cancelation adds another layer of complexity on top of this. I'll see whether I can come up with a simpler implementation.

Reply:
I think we can avoid using epoll though: xenopsd already has 1 thread/request, and there is nothing else to do if the file doesn't exist (it is not lke Lwt where it can go and execute something else), so doing a blocking Inotify.read should actually work and be simpler.